rust/src/arrow/record_batch.rs

```diff
@@ -18,26 +18,35 @@
 use arrow_array::{RecordBatch, RecordBatchReader};
 use arrow_schema::{ArrowError, SchemaRef};
 
-use crate::Result;
+use crate::{Error, Result};
 
 /// RecordBatchBuffer is a in-memory buffer for multiple [`RecordBatch`]s.
 ///
 ///
 #[derive(Debug)]
 pub struct RecordBatchBuffer {
```

Comment:
I'm wondering if we even want to keep this. It seems like all the uses can be handled either with `Vec<RecordBatch>` or `RecordBatchIterator<Vec<RecordBatch>::Iter>`. What do we think of removing it?

https://docs.rs/arrow-array/42.0.0/arrow_array/struct.RecordBatchIterator.html

Reply:
RecordBatchBuffer provides the ability to extend the record batches. RecordBatchIterator being an iterator (or wrapping an iterator) will only provide the read only view to the batches hence cannot do that. This functionality is used while writing batches limited by max rows per group in (dataset.rs::402, fragment.rs::84).
Vec<RecordBatch> provides that functionality but we cannot store schema for empty vector then.
Hence I think RecordBatchBuffer is best of both worlds so removing it is not feasible unless we change the implementation of fragment.rs::create and dataset.rs::write.
Please correct me if I misinterpreted anything. 
Thanks

Reply:
But are there any place where we are adding batches at the same time as we are iterating?

If we are adding batches, I think we can just use `Vec<RecordBatch>`.
If we are iterating, we could just use `RecordBatchIterator<>`.

Is there a place where we need both at the same time?

Reply:
Well, you're right, thanks for pointing this out. After skimming through the critical places it doesn't seem like we need both functionality at the same time. I'll start making the changes accordingly.